### PR TITLE
better case list form vlidation

### PR DESCRIPTION
**Pull request by snopoke**
http://manage.dimagi.com/default.asp?183624

@millerdev

### CODE REVIEW

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -2767,10 +2767,21 @@ def validate_for_build(self):
...
             case_tag = None
+            loaded_case_types = None
             for form in forms:
                 info = self.get_module_info()
                 form_info = {"id": form.id if hasattr(form, 'id') else None, "name": form.name}
                 non_auto_select_actions = [a for a in form.actions.load_update_cases if not a.auto_select]
+                this_forms_loaded_case_types = {action.case_type for action in non_auto_select_actions}
+                if loaded_case_types is None:
+                    loaded_case_types = this_forms_loaded_case_types
+                elif loaded_case_types != this_forms_loaded_case_types:
```
Comment (dannyroberts) [inline corehq/apps/app_manager/models.py]: This seems stricter than we need it to be (though that may be ok). My impression is that as long as (1) they all load case list screen defined by that module (the one that has the case list form) and (2) the path leading up to that screen is the same, then it'll build. (I'm not even totally certain we need (2)). Is that right? I'm not sure I understand this super well, but that was my impression looking at the error a few days ago.

**Comment by dannyroberts** [comment]
Looks good pending tests and small comment

File: corehq/apps/app_manager/models.py (hunk 1 above)
Comment (snopoke) [inline corehq/apps/app_manager/models.py]: yes, both 1 & 2 are correct but harder to test for. I'm fine leaving it as is until someone complains.

**Comment by millerdev** [comment]
GTG once tests pass. Speaking of tests, would be nice to have a test for the bug that this fixes.

File: corehq/apps/app_manager/tests/test_build_errors.py [hunk 2]
```diff
@@ -59,3 +60,25 @@ def test_case_tile_configuration_errors(self):
...
+        factory = AppFactory(build_version='2.11')
+        m0, m0f0 = factory.new_basic_module('register', 'person')
+        factory.form_opens_case(m0f0)
+
+        m1, m1f0 = factory.new_advanced_module('update', 'person', case_list_form=m0f0)
+        factory.form_requires_case(m1f0, case_type='house')
+        factory.form_requires_case(m1f0, parent_case_type='house')
+
+        m1f1 = factory.new_form(m1)
+        factory.form_requires_case(m1f1) # only loads a person case and not a house case
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_build_errors.py]: E261 at least two spaces before inline comment

File: corehq/apps/app_manager/tests/test_build_errors.py [hunk 3]
```diff
@@ -59,3 +60,25 @@ def test_case_tile_configuration_errors(self):
...
+
+        errors = factory.app.validate_app()
+        self.assertIn(case_tile_error, errors)
+
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_build_errors.py]: W391 blank line at end of file

File: corehq/apps/app_manager/tests/test_build_errors.py [hunk 4]
```diff
@@ -59,3 +60,25 @@ def test_case_tile_configuration_errors(self):
...
+        m1, m1f0 = factory.new_advanced_module('update', 'person', case_list_form=m0f0)
+        factory.form_requires_case(m1f0, case_type='house')
+        factory.form_requires_case(m1f0, parent_case_type='house')
+
+        m1f1 = factory.new_form(m1)
+        factory.form_requires_case(m1f1)  # only loads a person case and not a house case
+
+        errors = factory.app.validate_app()
+        self.assertIn(case_tile_error, errors)
+
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_build_errors.py]: W391 blank line at end of file

File: corehq/apps/app_manager/tests/test_build_errors.py [hunk 5]
```diff
@@ -59,3 +60,24 @@ def test_case_tile_configuration_errors(self):
             app = Application.wrap(source)
             errors = app.validate_app()
             self.assertIn(case_tile_error, errors)
+
+    def test_case_list_form_advanced_module_different_case_config(self):
```
Comment (snopoke) [inline corehq/apps/app_manager/tests/test_build_errors.py]: @dannyroberts :innocent:

Comment (dannyroberts) [inline corehq/apps/app_manager/tests/test_build_errors.py]: @millerdev I think this one's for you

**Comment by dannyroberts** [comment]
@snopoke looks like it can't be merged, but otherwise, this looks great and good to merge

**Comment by snopoke** [comment]
merged with master

**Comment by snopoke** [comment]
Cancelled previous build

File: corehq/apps/app_manager/tests/test_form_workflow.py [hunk 6]
```diff
@@ -214,6 +214,32 @@ def test_link_to_form_in_parent_module(self):
 
         self.assertXmlPartialEqual(self.get_xml('form_link_child_modules'), factory.app.create_suite(), "./entry[3]")
 
+    def test_form_links_submodule(self):
+        """
+        Test that when linking between two forms in a submodule we match up the session variables between the source
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_form_workflow.py]: E501 line too long (116 > 115 characters)

**Comment by millerdev** [comment]
Test failed. My guess is it's not real, but I'm not sure.
```
======================================================================
ERROR: setUpClass (custom.ewsghana.tests.ews_test_reminders.EWSTestReminders)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 138, in new_setUpClass
    result = original_setUpClass(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 138, in new_setUpClass
    result = original_setUpClass(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/custom/ewsghana/tests/ews_test_reminders.py", line 17, in setUpClass
    cls.sms_backend_mapping, cls.backend = create_backend()
  File "/home/travis/build/dimagi/commcare-hq/custom/ewsghana/utils.py", line 140, in create_backend
    backend.save()
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/couchdbkit/schema/base.py", line 134, in save
    db.save_doc(doc, **params)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/couchdbkit/client.py", line 515, in save_doc
    **params).json_body
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/restkit/resource.py", line 153, in put
    headers=headers, params_dict=params_dict, **params)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/couchdbkit/resource.py", line 140, in request
    response=e.response)
ResourceConflict: Document update conflict.
----------------------------------------------------------------------
```

**Comment by benrudolph** [comment]
This is not real see: https://github.com/dimagi/commcare-hq/pull/8596

**Comment by millerdev** [comment]
Since we're using `--failfast` when running tests, that likely means we did not get a complete test run here. What's the normal process?

1. Merge master into this branch and run tests again ... OR
2. Merge this PR hoping all is good.

**Comment by benrudolph** [comment]
Option 1 for sure, but right now I don't know how we restart the travis build on the merge commit

**Comment by snopoke** [comment]
:white_check_mark: :joy_cat:

**Pull request merged**